File: rust/src/83.remove-duplicates-from-sorted-list.rs
use crate::Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

#[allow(dead_code)]
impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[allow(dead_code)]
impl Solution {
    pub fn delete_duplicates(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut cur = &mut head;

        while let Some(current) = cur.as_mut() {
            while let Some(next) = &mut current.next {
                if current.val != next.val {
                    break;
                }

                current.next = next.next.take();
            }

            cur = &mut current.next;
        }

        head
    }
}

File: scala/83.remove-duplicates-from-sorted-list.scala
package leetcode.`83`

class ListNode(_x: Int = 0, _next: ListNode = null) {
  var next: ListNode = _next
  var x: Int = _x
}

object Solution {
  def deleteDuplicates(head: ListNode): ListNode = {
    @annotation.tailrec
    def recursion(prev: ListNode, current: ListNode): ListNode = {
      if (current == null) head
      else if (prev != null && prev.x == current.x) {
        prev.next = current.next
        recursion(prev, current.next)
      } else {
        recursion(current, current.next)
      }
    }

    recursion(null, head)
  }
}
